### PR TITLE
Add Keldysh ionization model

### PR DESCRIPTION
**Pull request by n01r**
According to eq. (9) in [D. Bauer and P. Mulser. Exact field ionization rates in the barrier-suppression regime
from numerical time-dependent Schrödinger-equation calculations.](http://journals.aps.org/pra/abstract/10.1103/PhysRevA.59.569)

Wiki entries will follow.

Depends on ~~#1541~~ and ~~#1542~~.

### CODE REVIEW

**Comment by ax3l** [comment]
@n01r pls rebase :)

**Comment by n01r** [comment]
@ax3l done

File: examples/LaserWakefield/include/simulation_defines/param/speciesDefinition.param [hunk 1]
```diff
@@ -143,6 +143,7 @@ struct Hydrogen
  *               -> linearly polarized lasers
  * - ADKCircPol : Ammosov-Delone-Krainov tunneling ionization (H-like)
  *                -> circularly polarized lasers
+ * - Keldysh : Keldysh ionization model
```
Comment (ax3l) [inline examples/LaserWakefield/include/simulation_defines/param/speciesDefinition.param]: @n01r pls update the documentation of those files, too:

```
./examples/SingleParticleTest/include/simulation_defines/param/speciesDefinition.param
./examples/WeibelTransverse/include/simulation_defines/param/speciesDefinition.param
./src/picongpu/include/simulation_defines/param/speciesDefinition.para
```

**Comment by n01r** [comment]
When you compile this (actually even with #1542) with CUDA 7.0 you'll get a warning that 
"_The ABI for passing parameters with 32-byte alignment has changed in GCC 4.6_"

Probably this is because some functor is passed by copy instead of by reference in the kernel call of `kernelIonizeParticles`. That can be treated in a follow-up where I can actually use @Heikman s generalization of the particle creation which is fairly identical to the kernel, anyway.

**Comment by ax3l** [comment]
we had that problem before in #36 
@psychocoderHPC ?

**Comment by psychocoderHPC** [comment]
Please post the full warning message that we can look to the line which creates the issue.
The solution is to give huge functors (data structs) by const reference to the function with the warning.

File: src/picongpu/include/particles/ionization/byField/Keldysh/AlgorithmKeldysh.hpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+         * \tparam BType type of magnetic field
+         * \tparam ParticleType type of particle to be ionized
+         *
+         * \param bField magnetic field value at t=0
+         * \param eField electric field value at t=0
+         * \param parentIon particle instance to be ionized with position at t=0 and momentum at t=-1/2
+         */
+        template<typename EType, typename BType, typename ParticleType >
+        HDINLINE void
+        operator()( const BType bField, const EType eField, ParticleType& parentIon, float_X randNr )
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/Keldysh/AlgorithmKeldysh.hpp]: bfield and eField must be a reference. It is not a native build in type.

**Comment by n01r** [comment]
/home/garten70/src/picongpu/src/picongpu/include/particles/ionization/ionization.hpp: In function ‘void picongpu::particles::ionization::kernelIonizeParticles(ParBoxIons, ParBoxElectrons, 
FrameIonizer, Mapping) [with ParBoxIons = PMacc::ParticlesBox<PMacc::Frame<PMacc::ParticlesBuffer<PMacc::ParticleDescription<boost::mpl::string<105, 0, 0, 0, 0, 0, 0, 0>, P
Macc::math::CT::Vector<mpl_::integral_c<int, 8>, mpl_::integral_c<int, 8>, mpl_::integral_c<int, 4> >, boost::mpl::v_item<pic
ongpu::placeholder_definition26::boundElectrons..... 
...
... loads and loads of unravelled templates of particle species
...
/home/garten70/src/picongpu/src/picongpu/include/particles/ionization/ionization.hpp:63:1: note: The ABI for passing parameters with 32-byte alignment has changed in
 GCC 4.6

File: examples/LaserWakefield/include/simulation_defines/param/speciesDefinition.param (hunk 1 above)
Comment (n01r) [inline examples/LaserWakefield/include/simulation_defines/param/speciesDefinition.param]: done

**Comment by ax3l** [comment]
@n01r did this already fix the gcc warning about the alignment or are there more places left?

**Comment by ax3l** [comment]
@n01r as discussed offline, it would be great if you could add the new model to the comparison examples (LWFA, droplet) that we did last year for the BSI & ADK models.

File: src/picongpu/include/particles/ionization/byField/Keldysh/Keldysh_Impl.hpp [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+        BlockArea BlockDescription;
+
+        private:
+
+            /* define ionization ALGORITHM (calculation) for ionization MODEL */
+            typedef T_IonizationAlgorithm IonizationAlgorithm;
+
+            /* random number generator */
+            typedef PMacc::random::RNGProvider<simDim, PMacc::random::methods::XorMin> RNGFactory;
+            typedef PMacc::random::distributions::Uniform<float> Distribution;
```
Comment (psychocoderHPC) [inline src/picongpu/include/particles/ionization/byField/Keldysh/Keldysh_Impl.hpp]: Must be an `float_X`

Comment (ax3l) [inline src/picongpu/include/particles/ionization/byField/Keldysh/Keldysh_Impl.hpp]: copied from `particles/synchrotronPhotons/PhotonCreator.hpp` and `particles/ionization/byField/ADK/ADK_Impl.hpp`

Comment (n01r) [inline src/picongpu/include/particles/ionization/byField/Keldysh/Keldysh_Impl.hpp]: :bow: missed that

**Pull request merged**